daml-assistant/daml-project-config/DA/Daml/Project/Config.hs

```diff
@@ -63,6 +76,85 @@ readConfig name path = do
     configE <- Y.decodeFileEither path
     fromRightM (throwIO . ConfigFileInvalid name) configE
 
+-- Substring for text (inclusive start, exclusive end)
+textSub :: Int -> Int -> Text -> Text
+textSub start end = T.take (end - start) . T.drop start
+
+data SubMatch = SubMatch
+  { smText :: Text
+  , smStart :: Int
+  , smEnd :: Int
+  , smLength :: Int
+  }
+
+-- TDFA has you working with less than ideal data types
+-- transform them into something more friendly
+transformSubmatch :: Array.Array Int (MatchText Text) -> [[SubMatch]]
+transformSubmatch = fmap (fmap toSubMatch . Array.elems) . Array.elems
+  where
+    toSubMatch :: (Text, (Int, Int)) -> SubMatch
+    toSubMatch (smText, (smStart, smLength)) = let smEnd = smStart + smLength in SubMatch {..}
+
+-- Finds any ${...} where the dollar isn't preceeded by (an odd number of) '\'
+-- Replaces the inner name with the value provided by the below mapping
+-- Note that `.` in the name are replaced with `_`, to allow for future support for heirarchical names
+interpolateEnvVariables :: [(String, String)] -> T.Text -> Either String T.Text
+interpolateEnvVariables env str = do
+  let matchPositions :: [[SubMatch]] = transformSubmatch $ getAllTextMatches (str =~ ("(^|[^\\\\])(\\\\*)\\${([^}]+)}" :: String))
+      -- First 2 matches aren't needed (whole match + start of line check)
+      --   TDFA doesn't support non capturing groups, which is why that match exists in the first place
+      replaceWithPrefix (s, start) [_, _, escapeSlashes, name] = do
```

Comment:
This one is a bit chewy to read - a type signature here would be helpful, as well as a more descriptive name for `s`.

Reply:
Ah right, I might pull it out into a `replaceAllInM :: Monad m => Regex -> String -> ([SubMatch] -> m String) -> m String`

Reply:
Did above, reads a lot better now IMO